doc/performance.md

```diff
@@ -1,39 +1,257 @@
 # Performance
 
-There is a Vagrantfile with performance inspection tools like SystemTap.
+- [**General guidelines**](#general-guidelines)
+- [**Default caching**](#default-caching)
+- [**Async reporting threads**](#async-reporting-threads)
+- [**Batching policy**](#batching-policy)
 
-```shell
-vagrant up
-vagrant ssh
-```
+This document gives some general guidelines to debug performance issues in
+APIcast. It also introduces the caching modes available and explains how they
+can help in increasing performance.
 
-APIcast is mounted into `/home/vagrant/app` so you can start it by:
+## General guidelines
 
-```shell
-cd app
-bin/apicast
-```
+When experiencing performance issues in APIcast, the first thing that we
+recommend to do is to try to determine the component that is responsible for
+those issues. In a typical APIcast deployment, there are three components to
+take into account:
 
-For profiling with stapxx it is recommended to start just one process and worker:
+- APIcast
+- The 3scale backend that authorizes requests and keeps track of the usage
+- The upstream API
 
-```shell
-bin/apicast -c examples/configuration/echo.json  > /dev/null
-```
+In order to measure the latency that APIcast plus the 3scale backend introduce,
+the first thing that we need to do is to measure the latency of the upstream
+API. There are many tools that can be used for that:
+[wrk](https://github.com/wg/wrk), [vegeta](https://github.com/tsenart/vegeta),
+[JMeter](https://jmeter.apache.org/), etc. The next step consists of using the
+same tool to run the same benchmark, but this time, pointing to APIcast instead
+of pointing to the upstream API directly. Comparing those results will give us
+an idea of the latency introduced by APIcast and the 3scale backend.
+
+If you see that the latency introduced by APIcast and the 3scale backend is too
+high, and you are using a self-managed APIcast that authorizes against the
+3scale backend SaaS, the first thing you can do is to make a simple request to
+the 3scale backend from the same machine where APIcast is deployed and measure
+the latency. Backend exposes a simple endpoint that just returns the version:
+https://su1.3scale.net/status. An authorization call requires more work because
+it needs to verify some keys, limits, and enqueue background jobs, among other
+things. The 3scale backend is very fast, on average, it does that in a matter of
+a few milliseconds, but still, it's more work than just checking the version
+like the `/status` endpoint does. So, if a request to `/status` takes, let's say
+around 300 ms from your APIcast environment, an authorization is going to take a
+bit more than that for every request that is not cached.
+
+
+## Default caching
 
-Then by opening another terminal you can use vegeta to create traffic:
+For requests that are not cached, APIcast extracts the usage metrics from
```

Comment:
Perhaps you want to reorganize this paragraph into a list of steps:

1. APIcast extracts the usage metrics from matching mapping rules
2. APIcast sends those metrics plus the application credentials
to the 3scale backend.
3. The 3scale backend makes the following tasks:
3.1.  Checks the application keys, and that the reported usage of metrics is within the defined limits.
3.2.  Enqueues a background job to increase the usage of the metrics reported.
3.3. Responds to APIcast whether the request should be authorized or not.
If it should be, the request goes to the upstream. Note that in this case, the request does not arrive to the upstream until the 3scale backend responds.